### PR TITLE
Xnor/patcher control

### PR DESCRIPTION
**Pull request by x37v**
I'd love if it looked a little nicer but, this gets the functionality into the runner panel in support for rnbo `1.2`

### CODE REVIEW

File: src/actions/device.ts [hunk 1]
```diff
@@ -93,12 +94,44 @@ export const savePresetToRemote = (name: string): AppThunk =>
 		oscQueryBridge.sendPacket(writePacket(message));
 	};
 
+export const loadPatcher = (name: string, inst: number = 0): AppThunk =>
+	() => {
+			let message = {
```
Comment (starakaj) [inline src/actions/device.ts:99]: whitespace?

File: src/actions/device.ts [hunk 2]
```diff
@@ -93,12 +94,44 @@ export const savePresetToRemote = (name: string): AppThunk =>
...
+					address: "/rnbo/inst/control/unload",
+					args: [
+						{ type: "i", value: inst }
+					]
+				};
+			}
+			oscQueryBridge.sendPacket(writePacket(message));
+		};
+
+export const unloadPatcher = (inst: number = 0): AppThunk =>
```
Comment (starakaj) [inline src/actions/device.ts:117]: It doesn't look like this is ever actually used? Why do we have UNLOAD_PATCHER_NAME and unloadPatcher? Seems like they do the same thing?

File: src/actions/device.ts [hunk 3]
```diff
@@ -115,7 +148,42 @@ export const initializeDevice = (desc: AnyJson): AppThunk =>
...
+		}
+	};
 
+export const initializePatchers = (desc: AnyJson): AppThunk =>
+	(dispatch, getState) => {
+		try {
+			const patcherDescriptions = (desc as any).CONTENTS || {};
+			let loadedName: string | undefined;
+
+			getPatchers(getState()).map((p) => {
```
Comment (starakaj) [inline src/actions/device.ts:162]: maybe reduce not map? I'm nitpicking though

File: src/actions/device.ts [hunk 4]
```diff
@@ -115,7 +148,42 @@ export const initializeDevice = (desc: AnyJson): AppThunk =>
...
+			console.log(e);
+		}
+	};
+
+export const setSelectedPatcher = (name: string): AppThunk =>
+	(dispatch, getState) => {
+		const state = getState();
+		try {
+			const updated: PatcherRecord[] = [];
+			getPatchers(state).map((p) => {
```
Comment (starakaj) [inline src/actions/device.ts:183]: again, maybe not map, but it's also fine

File: src/selectors/entities.ts [hunk 5]
```diff
@@ -35,3 +37,22 @@ export const getPreset = (state: RootStateType, id: string): PresetRecord | unde
...
+export const getPatchers = (state: RootStateType): EntityMap<PatcherRecord> => {
+	return state.entities[EntityType.PatcherRecord];
+};
+
+export const getPatcher = (state: RootStateType, name: string): PatcherRecord | undefined => {
+	return getEntity(state, EntityType.PatcherRecord, name);
+};
+
+export const getLoadedPatcher = (state: RootStateType): PatcherRecord | undefined => {
+	//there has to be a better way
```
Comment (starakaj) [inline src/selectors/entities.ts:50]: reduce!

**Review by starakaj — CHANGES_REQUESTED** [review]
Just a couple small things. There are some places where you're using map in a context where a forEach or a reduce seem more idiomatic. It of course works fine to use map, but maybe it's clearer to use an array function whose purpose is closer to what the code is trying to do. The other thing is that UNLOAD_PATCHER_NAME seems redundant with unloadPatcher, so one or the other should probably be removed.

File: src/actions/device.ts (hunk 1 above)
Comment (x37v) [reply src/actions/device.ts:99]: we use tabs so these show up weird on github, maybe there is some rendering setting we can do for the repo

**Comment by fde31** [comment]
Seems like I missed that a review was requested from me @x37v ? Oooops. Sorry! Anything u guys need at this point, anything that would be helpful from me - just holler!

**Comment by x37v** [comment]
I'd love it if someone could help make this new little widget look good

**Pull request merged**